libnetwork/iptables/firewalld.go

```diff
@@ -232,14 +247,36 @@ func setupDockerZone() error {
 		target:      "ACCEPT",
 	}
 	if err := connection.sysConfObj.Call(dbusInterface+".config.addZone", 0, dockerZone, dz.settings()).Err; err != nil {
-		return err
-	}
-	// Reload for change to take effect
-	if err := connection.sysObj.Call(dbusInterface+".reload", 0).Err; err != nil {
-		return err
+		return false, err
 	}
+	return true, nil
+}
 
-	return nil
+// setupDockerForwardingPolicy creates a policy to allow forwarding to anywhere to the docker
+// zone (where packets will be dealt with by docker's usual/non-firewalld configuration).
+// The bool return value is true if a firewalld reload is required.
+func setupDockerForwardingPolicy() (bool, error) {
+	// https://firewalld.org/documentation/man-pages/firewalld.dbus.html#FirewallD1.config
+	policy := map[string]interface{}{
+		"version":       "1.0",
+		"description":   "allow forwarding to the docker zone",
+		"ingress_zones": []string{"ANY"},
+		"egress_zones":  []string{dockerZone},
+		"target":        "ACCEPT",
+	}
+	if err := connection.sysConfObj.Call(dbusInterface+".config.addPolicy", 0, dockerFwdPolicy, policy).Err; err != nil {
+		if strings.HasPrefix(err.Error(), "NAME_CONFLICT") {
+			log.G(context.TODO()).Debugf("Firewalld: %s policy already exists", dockerFwdPolicy)
+			return false, nil
+		}
+		if strings.Contains(err.Error(), "UnknownMethod") {
+			log.G(context.TODO()).Debugf("Firewalld: addPolicy %s: unknown method", dockerFwdPolicy)
+			return false, nil
+		}
```

Comment:
The go-dbus docs for `(dbus.Call).Err` say that an error message from the peer will be of type `dbus.Error`. A `dbus.Error` has a `Name` value [which signifies the kind of error.](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names-error)
```suggestion
		var derr dbus.Error
		if errors.As(err, &derr) {
			if derr.Name == "NAME_CONFLICT" /* ??? */ {
				log.G(context.TODO()).Debugf("Firewalld: %s policy already exists", dockerFwdPolicy)
				return false, nil
			}
			if derr.Name == dbus.ErrMsgUnknownMethod.Name {
				log.G(context.TODO()).Debugf("Firewalld: addPolicy %s: unknown method", dockerFwdPolicy)
				return false, nil
			}
		}
```

Not sure what the deal is with firewalld's errors, though. They aren't documented to be namespaced like the d-bus spec says they should be.

Reply:
Right, yes - I've changed it to check explicitly for `Name == "org.freedesktop.DBus.Error.UnknownMethod"` (and checked that CentOS 7's ancient firewalld uses the same name).

But I don't think there's any way to do better for the firewalld error, it's just "an exception" ...

```
{
    Name: "org.fedoraproject.FirewallD1.Exception",
    Body: []interface{"NAME_CONFLICT: new_policy_object(): 'docker-forwarding'"},
}
```

Reply:
You could check that it's a firewalld exception before testing the first argument of the body?

Reply:
Ok - I realised I should have used errors.As() anyway, so I've made that change.

(Although I don't think it's likely something else would respond with "NAME_CONFLICT" and mean some other name conflict. If firewalld ever changes its error's Name, just checking the string might not break, but that's very unlikely too. And, whatever goes wrong here, it's only writing a log line - none of these error checks or returns change any other behaviour.)